src/qforte/utils/qbt_tapering.py

```diff
@@ -0,0 +1,321 @@
+                    idx += 1
+                    break
+        if idx == tau.shape[0]:
+            break
+    ####
+
+    # 4) Construct unitary matrix that will transform Hamiltonian
+
+    ## Translate binary vectors back to Pauli strings to obtain the generators of the symmetry group
+    gnrtrs = []
+    for i in range(tau.shape[0]):
+        pauli = qf.Circuit()
+        for j in reversed(range(n_qubits)):
+            if tau[i, j] != 0 or tau[i, j + n_qubits] != 0:
+                if tau[i, j] == 1 and tau[i, j + n_qubits] == 1:
+                    gate_type = 'Y'
+                elif tau[i, j] == 1:
+                    gate_type = 'X'
+                elif tau[i, j + n_qubits] == 1:
+                    gate_type = 'Z'
+                pauli.add_gate(qf.gate(gate_type, n_qubits - j - 1))
+        gnrtrs.append(pauli)
+
+    ## Find which sigma_x gate will be paired with which generator
+    sigma_x = np.zeros((tau.shape[0]), dtype=np.uint16)
+
+    if qiskit_order:
+        _range = range(n_qubits)
+    else:
+        _range = reversed(range(n_qubits))
+    for i in _range:
+        if np.argwhere(tau[:,n_qubits + i]).shape[0] == 1:
+            sigma_x[np.argwhere(tau[:,n_qubits + i])] = n_qubits - i - 1
+
+    ## Reorder sigma_x operators in descending order of qubit identity. Do the same with the generators for consistency.
+
+    reorder = np.argsort(sigma_x)
+    sigma_x = sigma_x[reorder][::-1]
+    gnrtrs = [gnrtrs[i] for i in reorder[::-1]]
+
+    ## Construct individual unitary matrices
+    untrs = []
+
+    for idx, generator in enumerate(gnrtrs):
+        circ = qf.Circuit()
+        circ.add_gate(qf.gate('X', sigma_x[idx]))
+        oprtr = qf.QubitOperator()
+        oprtr.add_term(1/np.sqrt(2), generator)
+        oprtr.add_term(1/np.sqrt(2), circ)
+        untrs.append(oprtr)
+
+    ## Construct final unitary matrix
+    unitary = qf.QubitOperator()
+    unitary.add_op(untrs[0])
+    for op in untrs[1:]:
+        unitary.operator_product(op, True, True)
+
+    # 5) Compute the various signs corresponding to the eigenvalues of the sigma_x operators that will be removed
+
+    signs = list(product([1, -1], repeat=len(sigma_x)))
+
+    return gnrtrs, sigma_x, untrs, unitary, signs
+
+def qubit_tapering_oprtr(tapered_qubits, sign, operator, unitary):
+    """
+    This function uses the unitary matrix obtained using find_Z2_symmetries to transform a given operator.
+    After the unitary transformation, the resulting operator acts on the qubits to be tapered off by at most
+    a single sigma_x operator. The sigma_x gates acting on qubits to be tapered off are replaced by their
+    +/- 1 eigenvalues, which are provided by the user.
+
+    Arguments
+
+    tapered_qubits: list of integers
+        The ID numbers of the qubits to be tapered off.
+
+    sign: list of +/- 1.
+        The desired sign structure corresponding to the eigenvalues of the sigma_x operators acting on
```

Comment:
"sign structure corresponding" is vague.

I think you mean _For each qubit to be tapered, should we assume we are in the + or - eigenspace? Takes one argument for each._